File: pset2/vigenere.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>

int getValue(char ch)
{
    if (ch >= 97 && ch <= 122)
    {
        return ch - 'a'; // returning value of character in range 0 to 25
    }
    else if (ch >= 65 && ch <= 90)
    {
        return ch - 'A'; // returning value of character in range 0 to 25
    }
    return ch;
}
int check(string arguement)
{
    for (int ch = 0; arguement[ch] != '\0'; ch++)
    {
        if (!isalpha(arguement[ch]))
        {
            return 0;
        }
    }
    return 1;
}
int main(int argc, string argv[])
{
    if (argc != 2 || atoi(argv[1]) || !check(argv[1])) // if arguement count is not 2 or arguement[1] is integer
    {
        printf("Usage: ./vigenere keyword\n");
        return 1;
    }
    else
    {
        string keyword = argv[1]; // getting keyword 
        int length_keyword = strlen(keyword);
        string plain_text = get_string("plaintext:  "); // getting plaintext
        for (int ch = 0, k = 0; plain_text[ch] != '\0'; ch++) // iterating over the plaintext
        {
            if (plain_text[ch] >= 65 && plain_text[ch] <= 90) // if plaintext + ch is upper case letter
            {
                plain_text[ch] = (plain_text[ch] - 'A' + getValue(keyword[k])) % 26 + 'A';
                k = (k + 1) % length_keyword; // modulus increment of keyword iterator
            }
            else if (plain_text[ch] >= 97 && plain_text[ch] <= 122) 
                // if plaintext + ch is lower case letter
            {    
                plain_text[ch] = (plain_text[ch] - 'a' + getValue(keyword[k])) % 26 + 'a';
                k = (k + 1) % length_keyword; // modulus increment of keyword iterator
            }
             
        }
        printf("ciphertext: %s\n", plain_text); // printing cipher text
        return 0;
    }
}

File: pset6/vigenere/vigenere.py
import sys
from cs50 import get_string


def is_valid(k):
    for ch in k:
        if not ch.isalpha():
            return False
    return True


def main():
    if len(sys.argv) != 2 or not is_valid(sys.argv[1]):
        print("Usage: ./vigenere k")
        sys.exit(1)

    k = sys.argv[1]  # getting key
    plaintext = get_string("plaintext: ")  # getting plain text
    j = 0

    print("ciphertext: ", end="")

    for ch in plaintext:
        if not ch.isalpha():
            print(ch, end="")
            continue

        ascii_offset = 65 if ch.isupper() else 97  # offset is 65 if upper case and 97 if lowercase

        pi = ord(ch) - ascii_offset  # converting alphabet to 0-25
        kj = ord(k[j % len(k)].upper()) - 65  # converting key value to 0-25
        ci = (pi + kj) % 26  # adding the key and plain text
        j += 1

        print(chr(ci + ascii_offset), end="")  # converting the cipher index to text

    print()

    return 0


if __name__ == "__main__":
    main()